src/jit/codegenxarch.cpp

```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->info.retStructDesc.passedInRegisters)
```

Comment:
If checks on line 182 and 184 can be replaced by

if (compMethodReturnsMultiRegRetType()) ?

Note that this method is not handling structs returned in two registers.  In case of structs returned in a single registers, return type is normalized.